### ISSUE TITLE
Modified climb

### ISSUE THREAD

**Issue by ataras2** [LouisDesdoigts/dLux#317]
I've been interested in using CLIMB for more general mask design. Suggested changes

- variable oversampling factor
- pulled out from optical layer in case people want to optimize something else than opd
- explicit comment correcting the equation of the plane (algo in paper incorrectly states it as ax + by + cz = 0)
- option of adding a dc offset before downsampling - this can simulation erosion/dilation. Done in a coefficient scale invariant way.

Modified code below originally taken from dLuxToliman. It also shows the Basis split - if we are doing @LouisDesdoigts 's fast fourier basis this needs to be an abstract class to keep the speed/flexibility

```python
# %%
from abc import abstractmethod

from jax import Array
import jax
import jax.numpy as np

import jax.random as jr
import matplotlib.colors as mcolors
import matplotlib.pyplot as plt
from abcdLux.mft import mft_kernels
import dLux.utils as dlu
import zodiax as zdx


def realpack_to_exp_transform(N: int) -> Array:
    """
    Build T (N,N) such that:
      c_exp = T @ a_real
    with real-packed a_real = [DC, cos1, sin1, cos2, sin2, ...]
    and exp-packed c_exp = [0, +1, -1, +2, -2, ...].

    Vectorized (no Python loop over k).
    """
    T = np.zeros((N, N), dtype=complex)

    # DC maps directly
    T = T.at[0, 0].set(1.0)

    # Number of cos/sin harmonic pairs present (excluding Nyquist if even)
    if N % 2 == 0:
        n_harm = (N // 2) - 1
        has_nyq = True
    else:
        n_harm = (N - 1) // 2
        has_nyq = False

    if n_harm > 0:
        k = np.arange(1, n_harm + 1)

        # exp indices for +k and -k in [0, +1, -1, +2, -2, ...]
        ip = 2 * k - 1
        im = 2 * k

        # real indices for cosk and sink in [DC, cos1, sin1, cos2, sin2, ...]
        ic = 2 * k - 1
        is_ = 2 * k

        half = np.array(0.5, dtype=complex)
        jhalf = np.array(0.5j, dtype=complex)

        # c(+k) = 0.5*cosk  - 0.5j*sink
        T = T.at[ip, ic].set(half)
        T = T.at[ip, is_].set(-jhalf)

        # c(-k) = 0.5*cosk  + 0.5j*sink
        T = T.at[im, ic].set(half)
        T = T.at[im, is_].set(+jhalf)

    # Nyquist (even N): only cosine exists; sin term is identically zero on-grid
    if has_nyq:
        T = T.at[N - 1, N - 1].set(1.0)

    return T


def cached_real_mft_kernels(N: int, M: int):
    """
    Returns Kx_real, Ky_real such that:
      out ≈ (Ky_real @ C_real @ Kx_real.T).real
    with C_real real-packed coefficients (N,N) and output shape (M,M).
    """
    # 1) Build the exp-basis kernels using your existing machinery.
    #    You want x_in = k-indices, x_out = spatial coords.
    #    We'll define coordinate specs directly as 1D arrays.

    # exp basis frequency indices: [0, +1, -1, +2, -2, ...]
    # (same along x and y)
    def k_order_monotone(N):
        if N % 2 == 1:
            kmax = (N - 1) // 2
            ks = [0]
            for m in range(1, kmax + 1):
                ks.extend([m, -m])
        else:
            kmax = N // 2
            ks = [0]
            for m in range(1, kmax):
                ks.extend([m, -m])
            ks.append(kmax)  # Nyquist
        return np.asarray(ks)

    kx = k_order_monotone(N)
    ky = k_order_monotone(N)

    # centered output coordinates in cycles
    n = np.arange(M)
    x = (n - (M // 2)) / M
    y = (n - (M // 2)) / M

    alpha = 2 * np.pi
    weight = 1.0

    # Your builder expects spec_in/spec_out; pass explicit coords.
    # spec_in is kx/ky, spec_out is x/y.
    Kx, Ky = mft_kernels((kx, ky), (x, y), alpha=alpha, weight=weight)
    Kx = Kx.astype(complex)
    Ky = Ky.astype(complex)

    # 2) Build the real->exp coefficient transforms
    Tx = realpack_to_exp_transform(N)
    Ty = realpack_to_exp_transform(N)

    # 3) Fold transforms into kernels: Ky_real = Ky @ Ty, Kx_real = Kx @ Tx
    Kx_real = Kx @ Tx
    Ky_real = Ky @ Ty

    return Kx_real, Ky_real


def eval_fourier2d_real_fast(C_real: Array, Kx_real: Array, Ky_real: Array) -> Array:
    """
    Evaluate with cached real-output kernels.
    C_real: (N,N) real-packed coeffs (float)
    Kx_real: (M,N) complex
    Ky_real: (M,N) complex
    Returns: (M,M) real (float)
    """
    out = Ky_real @ (C_real.astype(Kx_real.dtype) @ Kx_real.T)
    return out.real


N = 10
M = 512


Kx_r, Ky_r = cached_real_mft_kernels(N, M)


def eval_demo(c):
    return eval_fourier2d_real_fast(c, Kx_r, Ky_r)


c = np.zeros((N, N))

c = c.at[0, 1].set(1.0)


plt.imshow(eval_demo(c))
plt.colorbar()

np.max(eval_demo(c))

# %%


class Basis(zdx.base.Base):

    @property
    def shape(self):
        pass

    @abstractmethod
    def eval(self, coeffs: Array) -> Array:
        pass


class DirectBasis(Basis):
    basis: Array

    def __init__(self, basis):
        self.basis = basis

    def eval(self, coeffs: Array) -> Array:
        return dlu.eval_basis(self.basis, coeffs)

    @property
    def shape(self):
        return self.basis.shape

    @property
    def coeffs_shape(self):
        return (len(self.basis),)


class ImplictFourierBasis(Basis):
    """
    uses a precomputed fourier basis to save memory
    """

    Kx_r: Array
    Ky_r: Array

    def __init__(self, N, M):
        self.Kx_r, self.Ky_r = cached_real_mft_kernels(N, M)

    def eval(self, coeffs: Array) -> Array:
        return eval_fourier2d_real_fast(coeffs, self.Kx_r, self.Ky_r)

    @property
    def shape(self):
        return self.Kx_r.shape[0], self.Ky_r.shape[0]

    @property
    def coeffs_shape(self):
        return self.Kx_r.shape[1], self.Ky_r.shape[1]


def rms(arr, mask=None):
    if mask is None:
        mask = np.ones_like(arr)
    return np.sqrt(np.mean((arr[mask > 0]) ** 2))


class BasisCLIMB(zdx.base.Base):
    """CLIMB binarization wrapper around a continuous wavefront `Basis`.

    The class evaluates a continuous field, optionally applies an RMS-scaled
    offset, tiles it into `n_oversample x n_oversample` cells, and computes a
    soft binary occupancy value per cell.
    """

    basis: Basis
    coeffs: Array

    n_oversample: int

    offset_rms: Array

    def __init__(self, basis: Basis, n_oversample: int = 3, offset_rms: float = 0.0):
        """Create a CLIMB parameterization.

        Args:
            basis: Basis used to evaluate the continuous field.
            n_oversample: Number of pixels per CLIMB cell side (> 1).
            offset_rms: Additive offset in units of field RMS. This has the effect of eroding (<0)
                or dilating (>0) the binary occupancy.
        """

        # Requirements:
        # - `n_oversample` is an integer > 1
        # - basis output shape is divisible by `n_oversample` in both axes
        if not isinstance(n_oversample, int):
            raise ValueError("n_oversample must be an integer greater than 1")
        if n_oversample <= 1:
            raise ValueError("n_oversample must be an integer greater than 1")

        if basis.shape[0] % n_oversample != 0 or basis.shape[1] % n_oversample != 0:
            raise ValueError("Basis output shape must be a multiple of n_oversample")

        self.basis = basis
        self.n_oversample = n_oversample
        self.coeffs = np.zeros(basis.coeffs_shape)

        self.offset_rms = np.array(offset_rms)

    def lsq_params(self, img: Array):
        """Return least-squares plane-fit parameters for an image patch.

        Fits $z(x,y)=ax+by+c$ to `img` over normalized coordinates in $[0,1]$.
        """
        xx, yy = np.meshgrid(
            np.linspace(0, 1, img.shape[0]), np.linspace(0, 1, img.shape[1])
        )
        A = np.vstack([xx.ravel(), yy.ravel(), np.ones_like(xx).ravel()]).T
        matrix = np.linalg.inv(np.dot(A.T, A)).dot(A.T)
        return matrix, xx, yy, A

    def lsq(self, img: Array):
        """Return plane coefficients `(a, b, c)` for `img`."""
        matrix, _, _, _ = self.lsq_params(img)
        return np.dot(matrix, img.ravel())

    def area(self, img: Array, epsilon: float = 1e-15):
        """Compute soft binary area occupancy for a patch.

        The patch is approximated with a plane and converted to an occupancy-like
        scalar in $[0,1]$.
        """
        a, b, c = self.lsq(img)
        a = np.where(a == 0, epsilon, a)
        b = np.where(b == 0, epsilon, b)
        c = np.where(c == 0, epsilon, c)
        x1 = (-b - c) / (a)  # don't divide by zero
        x2 = -c / (a)  # don't divide by zero
        x1, x2 = np.min(np.array([x1, x2])), np.max(np.array([x1, x2]))
        x1, x2 = np.max(np.array([x1, 0.0])), np.min(np.array([x2, 1.0]))

        area_fraction = (
            x1
            + (-c / b) * x2
            - (0.5 * a / b) * x2**2
            - (-c / b) * x1
            + (0.5 * a / b) * x1**2
        )

        # Set the regions where there is a defined gradient
        area_fraction = np.where(area_fraction >= 0.5, area_fraction, 1 - area_fraction)

        # Colour in regions
        area_fraction = np.where(np.mean(img) >= 0, area_fraction, 1 - area_fraction)

        # Force fully positive / fully non-positive patches to hard values.
        area_fraction = np.where(np.all(img > 0), 1, area_fraction)
        area_fraction = np.where(np.all(img <= 0), 0, area_fraction)

        # undecided region
        area_fraction = np.where(np.any(img == 0), area_fraction > 0, area_fraction)

        # rescale between 0 and 1
        area_fraction = np.clip(area_fraction, 0, 1)

        return area_fraction

    def CLIMB(self):
        """Evaluate the CLIMB soft-binary image from current coefficients."""
        n_cells = self.basis.shape[-1] // self.n_oversample
        block = self.n_oversample

        wf = self.basis.eval(self.coeffs)

        wf += self.offset_rms * rms(wf)

        # Tile into `(n_cells * n_cells)` blocks of shape `(block, block)`.
        # (n_cells, block, n_cells, block) -> (n_cells, n_cells, block, block)
        # -> (n_cells * n_cells, block, block)
        tiles = wf.reshape(n_cells, block, n_cells, block).transpose(0, 2, 1, 3)
        flat_tiles = tiles.reshape(-1, block, block)

        vmap_mask = jax.vmap(self.area, in_axes=(0))

        soft_bin = vmap_mask(flat_tiles).reshape(n_cells, n_cells)

        return soft_bin


N = 30
M = 512 * 3

basis = ImplictFourierBasis(N, M)

bc = BasisCLIMB(basis)

img_raw = bc.basis.eval(
    jr.normal(jr.key(0), basis.coeffs_shape)
    + 1j * jr.normal(jr.key(0), basis.coeffs_shape)
)

i, j = 11, 0

plt.imshow(img_raw[i : i + 3, j : j + 3], norm=mcolors.CenteredNorm(), cmap="bwr")
plt.colorbar()

bc.area(img_raw[i : i + 3, j : j + 3])

# %%


bc = BasisCLIMB(basis)


def climb_image(c, bc):
    return bc.set("coeffs", c).CLIMB()


img = climb_image(np.zeros(basis.coeffs_shape), bc)
img = climb_image(
    jr.normal(jr.key(0), basis.coeffs_shape)
    + 1j * jr.normal(jr.key(0), basis.coeffs_shape),
    bc,
)

plt.imshow(img)

img_2 = climb_image(
    jr.normal(jr.key(0), basis.coeffs_shape)
    + 1j * jr.normal(jr.key(0), basis.coeffs_shape),
    bc.set("offset_rms", 0.1),
)

plt.imshow(img_2 - img, norm=mcolors.CenteredNorm(), cmap="bwr")

# %%
plt.figure()
c = jr.normal(jr.key(0), basis.coeffs_shape) + 1j * jr.normal(
    jr.key(0), basis.coeffs_shape
)

for i, rms_v in enumerate([-0.1, 0.0, 0.1]):
    plt.subplot(1, 3, i + 1)

    img = climb_image(c, bc.set("offset_rms", rms_v))

    plt.imshow(img)
    plt.title(f"offset_rms = {rms_v}")

# %%
```

**Comment by benjaminpope** [comment]
Love to see it developed more than a literal back of the envelope

**Comment by LouisDesdoigts** [comment]
Yep I got a general version sitting on local in my present updates im working on

**Comment by LouisDesdoigts** [comment]
Completed in v0.15.0
